File: minipdf-python/src/minipdf/pdf.py
from __future__ import annotations

from dataclasses import dataclass, field


def _number(value: float) -> str:
    return f"{value:.3f}".rstrip("0").rstrip(".") or "0"


def _pdf_text(value: str) -> bytes:
    encoded = value.encode("cp1252", errors="replace")
    return encoded.replace(b"\\", b"\\\\").replace(b"(", b"\\(").replace(b")", b"\\)")


@dataclass(frozen=True, slots=True)
class TextStyle:
    size: float = 11.0
    bold: bool = False
    italic: bool = False

    @property
    def resource(self) -> str:
        if self.bold and self.italic:
            return "F4"
        if self.bold:
            return "F2"
        if self.italic:
            return "F3"
        return "F1"


@dataclass(slots=True)
class PdfPage:
    width: float
    height: float
    operations: list[bytes] = field(default_factory=list)

    def add_text(self, text: str, x: float, y: float, style: TextStyle) -> None:
        self.operations.append(
            b"BT /"
            + style.resource.encode("ascii")
            + b" "
            + _number(style.size).encode("ascii")
            + b" Tf 0 0 0 rg 1 0 0 1 "
            + _number(x).encode("ascii")
            + b" "
            + _number(y).encode("ascii")
            + b" Tm ("
            + _pdf_text(text)
            + b") Tj ET\n"
        )

    def content(self) -> bytes:
        return b"".join(self.operations)


@dataclass(slots=True)
class PdfDocument:
    pages: list[PdfPage] = field(default_factory=list)

    def add_page(self, width: float, height: float) -> PdfPage:
        page = PdfPage(width, height)
        self.pages.append(page)
        return page

    def to_bytes(self) -> bytes:
        if not self.pages:
            self.add_page(595.28, 841.89)

        page_count = len(self.pages)
        font_start = 3 + page_count * 2
        objects: list[bytes] = [
            b"<< /Type /Catalog /Pages 2 0 R >>",
            b"<< /Type /Pages /Kids ["
            + b" ".join(f"{3 + index * 2} 0 R".encode("ascii") for index in range(page_count))
            + f"] /Count {page_count} >>".encode("ascii"),
        ]

        for index, page in enumerate(self.pages):
            page_id = 3 + index * 2
            content_id = page_id + 1
            resources = (
                f"/Font << /F1 {font_start} 0 R /F2 {font_start + 1} 0 R "
                f"/F3 {font_start + 2} 0 R /F4 {font_start + 3} 0 R >>"
            )
            objects.append(
                (
                    f"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 {_number(page.width)} "
                    f"{_number(page.height)}] /Resources << {resources} >> "
                    f"/Contents {content_id} 0 R >>"
                ).encode("ascii")
            )
            content = page.content()
            objects.append(
                f"<< /Length {len(content)} >>\nstream\n".encode("ascii") + content + b"endstream"
            )

        for base_font in (
            "Helvetica",
            "Helvetica-Bold",
            "Helvetica-Oblique",
            "Helvetica-BoldOblique",
        ):
            font_object = (
                f"<< /Type /Font /Subtype /Type1 /BaseFont /{base_font} "
                "/Encoding /WinAnsiEncoding >>"
            )
            objects.append(font_object.encode("ascii"))

        output = bytearray(b"%PDF-1.4\n%\xe2\xe3\xcf\xd3\n")
        offsets = [0]
        for object_id, content in enumerate(objects, start=1):
            offsets.append(len(output))
            output.extend(f"{object_id} 0 obj\n".encode("ascii"))
            output.extend(content)
            output.extend(b"\nendobj\n")
        xref_offset = len(output)
        output.extend(f"xref\n0 {len(objects) + 1}\n".encode("ascii"))
        output.extend(b"0000000000 65535 f \n")
        for offset in offsets[1:]:
            output.extend(f"{offset:010d} 00000 n \n".encode("ascii"))
        output.extend(
            (
                f"trailer\n<< /Size {len(objects) + 1} /Root 1 0 R >>\n"
                f"startxref\n{xref_offset}\n%%EOF\n"
            ).encode("ascii")
        )
        return bytes(output)

File: minipdf-go/pdf.go
package minipdf

import (
	"bytes"
	"fmt"
	"strconv"
	"strings"
	"unicode"
)

type PDFColor struct {
	Red   float64
	Green float64
	Blue  float64
}

var (
	PDFColorBlack       = PDFColor{}
	PDFColorWhite       = PDFColor{Red: 1, Green: 1, Blue: 1}
	PDFColorLightGray   = PDFColor{Red: 0.92, Green: 0.92, Blue: 0.92}
	PDFColorTableHeader = PDFColor{Red: 0.86, Green: 0.91, Blue: 0.96}
)

type pdfOperation interface {
	appendPDF(*bytes.Buffer)
}

type PDFDocument struct {
	pages []*PDFPage
}

type PDFPage struct {
	Width      float64
	Height     float64
	operations []pdfOperation
}

func NewPDFDocument() *PDFDocument {
	return &PDFDocument{}
}

func (document *PDFDocument) AddPage(width, height float64) *PDFPage {
	page := &PDFPage{Width: width, Height: height}
	document.pages = append(document.pages, page)
	return page
}

func (page *PDFPage) AddText(text string, x, y, fontSize float64, color PDFColor, bold bool) {
	page.operations = append(page.operations, textOperation{
		text: text, x: x, y: y, fontSize: fontSize, color: color, bold: bold,
	})
}

func (page *PDFPage) AddRect(x, y, width, height float64, color PDFColor) {
	page.operations = append(page.operations, rectOperation{
		x: x, y: y, width: width, height: height, color: color,
	})
}

func (page *PDFPage) AddLine(x1, y1, x2, y2 float64, color PDFColor, width float64) {
	page.operations = append(page.operations, lineOperation{
		x1: x1, y1: y1, x2: x2, y2: y2, color: color, width: width,
	})
}

type textOperation struct {
	text           string
	x, y, fontSize float64
	color          PDFColor
	bold           bool
}

func (operation textOperation) appendPDF(buffer *bytes.Buffer) {
	font := "F1"
	if operation.bold {
		font = "F2"
	}
	fmt.Fprintf(buffer, "BT /%s %s Tf %s %s %s rg %s %s Td (%s) Tj ET\n",
		font,
		pdfNumber(operation.fontSize),
		pdfNumber(operation.color.Red),
		pdfNumber(operation.color.Green),
		pdfNumber(operation.color.Blue),
		pdfNumber(operation.x),
		pdfNumber(operation.y),
		escapePDFText(operation.text),
	)
}

type rectOperation struct {
	x, y, width, height float64
	color               PDFColor
}

func (operation rectOperation) appendPDF(buffer *bytes.Buffer) {
	fmt.Fprintf(buffer, "%s %s %s rg %s %s %s %s re f\n",
		pdfNumber(operation.color.Red),
		pdfNumber(operation.color.Green),
		pdfNumber(operation.color.Blue),
		pdfNumber(operation.x),
		pdfNumber(operation.y),
		pdfNumber(operation.width),
		pdfNumber(operation.height),
	)
}

type lineOperation struct {
	x1, y1, x2, y2 float64
	color          PDFColor
	width          float64
}

func (operation lineOperation) appendPDF(buffer *bytes.Buffer) {
	fmt.Fprintf(buffer, "%s %s %s RG %s w %s %s m %s %s l S\n",
		pdfNumber(operation.color.Red),
		pdfNumber(operation.color.Green),
		pdfNumber(operation.color.Blue),
		pdfNumber(operation.width),
		pdfNumber(operation.x1),
		pdfNumber(operation.y1),
		pdfNumber(operation.x2),
		pdfNumber(operation.y2),
	)
}

func (document *PDFDocument) Bytes() []byte {
	pages := document.pages
	if len(pages) == 0 {
		pages = []*PDFPage{{Width: PageSizeA4.Width, Height: PageSizeA4.Height}}
	}

	pageCount := len(pages)
	objects := make([][]byte, 4+pageCount*2)
	objects[0] = []byte("<< /Type /Catalog /Pages 2 0 R >>")

	pageReferences := make([]string, pageCount)
	for index := range pages {
		pageReferences[index] = fmt.Sprintf("%d 0 R", 5+index*2)
	}
	objects[1] = []byte(fmt.Sprintf("<< /Type /Pages /Count %d /Kids [%s] >>", pageCount, strings.Join(pageReferences, " ")))
	objects[2] = []byte("<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica /Encoding /WinAnsiEncoding >>")
	objects[3] = []byte("<< /Type /Font /Subtype /Type1 /BaseFont /Helvetica-Bold /Encoding /WinAnsiEncoding >>")

	for index, page := range pages {
		pageObjectNumber := 5 + index*2
		contentObjectNumber := pageObjectNumber + 1
		objects[pageObjectNumber-1] = []byte(fmt.Sprintf(
			"<< /Type /Page /Parent 2 0 R /MediaBox [0 0 %s %s] /Resources << /Font << /F1 3 0 R /F2 4 0 R >> >> /Contents %d 0 R >>",
			pdfNumber(page.Width), pdfNumber(page.Height), contentObjectNumber,
		))
		var content bytes.Buffer
		for _, operation := range page.operations {
			operation.appendPDF(&content)
		}
		objects[contentObjectNumber-1] = []byte(fmt.Sprintf("<< /Length %d >>\nstream\n%sendstream", content.Len(), content.String()))
	}

	var output bytes.Buffer
	output.WriteString("%PDF-1.4\n%\xe2\xe3\xcf\xd3\n")
	offsets := make([]int, len(objects)+1)
	for index, object := range objects {
		offsets[index+1] = output.Len()
		fmt.Fprintf(&output, "%d 0 obj\n", index+1)
		output.Write(object)
		output.WriteString("\nendobj\n")
	}
	xrefOffset := output.Len()
	fmt.Fprintf(&output, "xref\n0 %d\n0000000000 65535 f \n", len(objects)+1)
	for index := 1; index < len(offsets); index++ {
		fmt.Fprintf(&output, "%010d 00000 n \n", offsets[index])
	}
	fmt.Fprintf(&output, "trailer\n<< /Size %d /Root 1 0 R >>\nstartxref\n%d\n%%%%EOF\n", len(objects)+1, xrefOffset)
	return output.Bytes()
}

func pdfNumber(value float64) string {
	return strconv.FormatFloat(value, 'f', -1, 64)
}

func escapePDFText(text string) string {
	var escaped strings.Builder
	for _, character := range text {
		switch character {
		case '\\', '(', ')':
			escaped.WriteByte('\\')
			escaped.WriteRune(character)
		case '\n', '\r', '\t':
			escaped.WriteByte(' ')
		default:
			if character <= 0xff && !unicode.IsControl(character) {
				escaped.WriteRune(character)
			} else {
				escaped.WriteByte('?')
			}
		}
	}
	return escaped.String()
}
